# How to deal with categorical variable.

Categorical variables are variables/values/features in our dataset that are non-numerical. i.e. they are usually in text format (i.e. datatype is object). The opposite of categorical values in our dataset are called continuous variables. A continuous variable can be numeric or a date/time.

### There are two main types of categorical variables, these are Nominal and Ordinal.

1.	Nominal feature: these are feature where the categories are only labelled without any order of precedence. For example, a feature like gender having two categories (male and female) has no particular order.  

2.	Ordinal feature:  these are features which have some order associated with. For example, a feature like economic status, with three categories: low, medium and high, which have an order associated with them.


Now why are we concerned about categorical variables:

•	“Many machine learning models, such as regression or SVM, are algebraic. This means that their input must be numerical. To use these models, categories must be transformed into numbers first, before you can apply the learning algorithm on them."

•	“While some ML packages or libraries might transform categorical data to numeric automatically based on some default embedding method, many other ML packages don’t support such inputs.”


•	“The machine doesn’t interpret categorical data like the way we humans do. For example, let take the names of the cities: Abuja, Minna and Accra, humans tends to categorized the cities Abuja and Minna as one i.e. similar and Accra to be distinct from these two because of the relationship between Abuja and Minna is closer compare to any of them to Accra (Abuja and Minna are in the same country, Nigeria). But for the machine model these cities are just three different levels (possible values) of the same feature City. If you do not specify the additional contextual information, it will be impossible for the model to differentiate between highly different levels.”


Now because of all these reasons we are now face with the task of transforming these categorical data into numerical data for further processing and/or at the same time getting additional, useful information from these categorical columns which is one of the process in feature engineering.
In this class we will be using the Housing Price competition dataset from kaggle to explain all you need to know in handling categorical variables.



Let jump into our cell to communicate in codes

First we will start by importing the basic libraries we need 

In [2]:
import pandas as pd
import numpy as np
import copy

Now let read out our data into a pandas dataframe

In [3]:
df = pd.read_csv('housing_train.csv')

In [4]:
df.head()

Id  MSSubClass MSZoning  LotFrontage  LotArea Street Alley LotShape  \
0   1          60       RL         65.0     8450   Pave   NaN      Reg   
1   2          20       RL         80.0     9600   Pave   NaN      Reg   
2   3          60       RL         68.0    11250   Pave   NaN      IR1   
3   4          70       RL         60.0     9550   Pave   NaN      IR1   
4   5          60       RL         84.0    14260   Pave   NaN      IR1   

  LandContour Utilities  ... PoolArea PoolQC Fence MiscFeature MiscVal MoSold  \
0         Lvl    AllPub  ...        0    NaN   NaN         NaN       0      2   
1         Lvl    AllPub  ...        0    NaN   NaN         NaN       0      5   
2         Lvl    AllPub  ...        0    NaN   NaN         NaN       0      9   
3         Lvl    AllPub  ...        0    NaN   NaN         NaN       0      2   
4         Lvl    AllPub  ...        0    NaN   NaN         NaN       0     12   

  YrSold  SaleType  SaleCondition  SalePrice  
0   2008        WD         Normal     208500  
1   2007        WD         Normal     181500  
2   2008        WD         Normal     223500  
3   2006        WD        Abnorml     140000  
4   2008        WD         Normal     250000  

[5 rows x 81 columns]

We can see from the head of the dataframe there are some missing values (NaN) and there are integer, object, and float cloumns.

But let run some codes to get this information

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1460 entries, 0 to 1459
Data columns (total 81 columns):
Id               1460 non-null int64
MSSubClass       1460 non-null int64
MSZoning         1460 non-null object
LotFrontage      1201 non-null float64
LotArea          1460 non-null int64
Street           1460 non-null object
Alley            91 non-null object
LotShape         1460 non-null object
LandContour      1460 non-null object
Utilities        1460 non-null object
LotConfig        1460 non-null object
LandSlope        1460 non-null object
Neighborhood     1460 non-null object
Condition1       1460 non-null object
Condition2       1460 non-null object
BldgType         1460 non-null object
HouseStyle       1460 non-null object
OverallQual      1460 non-null int64
OverallCond      1460 non-null int64
YearBuilt        1460 non-null int64
YearRemodAdd     1460 non-null int64
RoofStyle        1460 non-null object
RoofMatl         1460 non-null object
Exterior1st      1460 non-n

Using the method .info() method gives us the basic information we need about the dataframe. 

As we can see we have 3 float columns, 35 integer columns and 43 object columns making it 81 total columns in total and we have a total of 1460 entries (rows).

   ### Note
entries is the same as rows.

columns is the same as features

Now because we are dealing with categorical variables we are concern with the dtype: object columns.

So we will be creating a new dataframe containing only the dtype: object columns.

Also we will be using the copy() method to copy dataframe in order not to tamper with the real dataframe since we are going to be doing a lot of analysis on different dataframes.

In [6]:
cat_df = df.select_dtypes(include=['object']).copy()

In [30]:
cat_df.head()

MSZoning Street  Alley LotShape LandContour Utilities LotConfig LandSlope  \
0       RL   Pave  SBrkr      Reg         Lvl    AllPub    Inside       Gtl   
1       RL   Pave  SBrkr      Reg         Lvl    AllPub       FR2       Gtl   
2       RL   Pave  SBrkr      IR1         Lvl    AllPub    Inside       Gtl   
3       RL   Pave  SBrkr      IR1         Lvl    AllPub    Corner       Gtl   
4       RL   Pave  SBrkr      IR1         Lvl    AllPub       FR2       Gtl   

  Neighborhood Condition1  ... GarageType GarageFinish GarageQual GarageCond  \
0      CollgCr       Norm  ...     Attchd          RFn         TA         TA   
1      Veenker      Feedr  ...     Attchd          RFn         TA         TA   
2      CollgCr       Norm  ...     Attchd          RFn         TA         TA   
3      Crawfor       Norm  ...     Detchd          Unf         TA         TA   
4      NoRidge       Norm  ...     Attchd          RFn         TA         TA   

  PavedDrive PoolQC  Fence MiscFeature SaleType SaleCondition  
0          Y  SBrkr  SBrkr       SBrkr       WD        Normal  
1          Y  SBrkr  SBrkr       SBrkr       WD        Normal  
2          Y  SBrkr  SBrkr       SBrkr       WD        Normal  
3          Y  SBrkr  SBrkr       SBrkr       WD       Abnorml  
4          Y  SBrkr  SBrkr       SBrkr       WD        Normal  

[5 rows x 43 columns]

Getting null values

In [8]:
cat_df.isnull().values.sum()

6617

Getting null values column-wise

In [9]:
 cat_df.isnull().sum()

MSZoning            0
Street              0
Alley            1369
LotShape            0
LandContour         0
Utilities           0
LotConfig           0
LandSlope           0
Neighborhood        0
Condition1          0
Condition2          0
BldgType            0
HouseStyle          0
RoofStyle           0
RoofMatl            0
Exterior1st         0
Exterior2nd         0
MasVnrType          8
ExterQual           0
ExterCond           0
Foundation          0
BsmtQual           37
BsmtCond           37
BsmtExposure       38
BsmtFinType1       37
BsmtFinType2       38
Heating             0
HeatingQC           0
CentralAir          0
Electrical          1
KitchenQual         0
Functional          0
FireplaceQu       690
GarageType         81
GarageFinish       81
GarageQual         81
GarageCond         81
PavedDrive          0
PoolQC           1453
Fence            1179
MiscFeature      1406
SaleType            0
SaleCondition       0
dtype: int64

This is the list of columns with missing values

In [10]:
null_list = ['Electrical',
'BsmtQual',        
'BsmtCond',           
'BsmtExposure',       
'BsmtFinType1',      
'BsmtFinType2',      
'Alley',    
'PoolQC',           
'Fence',          
'MiscFeature',      
'FireplaceQu',      
'GarageType',       
'GarageFinish',       
'GarageQual',      
'GarageCond' ]

 We are going to be replacing the missing values with the mode i.e the highest occuring value of each columns.

In [11]:
for i in null_list:
    cat_df = cat_df.fillna(cat_df[i].value_counts().index[0])

Basically the .value_counts() is printing out the number of times each value appears in its column, while the .index[0] method is picking out the first item in the series which happens to be the highest occurring value. The. fillna() method is replacing this highest occurring value with the missing value in its column.  

Let check if there are still null values

In [12]:
cat_df.isnull().values.sum()

0

As we can see there no missing values anymore.


### Note
There are various ways to fill missing values. You can decide to use their mean, median, mode, standard deviation or you can even decide to drop them.

## Ways to destroy categorical variables.

### We will be using six different techniques to encode the categorical variables in the Housing Price competition dataset.

Here is the list of the techniques we will be using:

1.	Replacing values
2.	Encoding labels
3.	One-Hot encoding
4.	Binary encoding
5.	Backward difference encoding
 


### Replacing values

This is the basic way of encoding categorical values. It is just by literrally asigning numerical values to the values in the categorical columns.
This can be easily acheive using pandas built in function replace().

### Note
For the sake of simplicity we will be making use of only the Condition1 column for this class.

We will create a dictionary of dictionary assigning the parent dictionary's key to be the name of the column we want to encode, while the parent ditionary's value to be the child dictionary. The child dictionary's keys will be  the column values and the child dictionary's values wiil be our fabricated values.
This is what I mean:

In [13]:
replace_dict = {'Condition1': {'Norm': 1, 'Feedr': 2, 'PosN': 3, 'Artery': 4,
                                  'RRAe': 5, 'RRNn': 6, 'RRAn': 7 , 'PosA': 8 , 'RRNe': 9}}


In [14]:
cat_df_replace = cat_df.copy()

In [15]:
cat_df_replace.replace(replace_dict, inplace=True)

In [16]:
cat_df_replace.head()

MSZoning Street  Alley LotShape LandContour Utilities LotConfig LandSlope  \
0       RL   Pave  SBrkr      Reg         Lvl    AllPub    Inside       Gtl   
1       RL   Pave  SBrkr      Reg         Lvl    AllPub       FR2       Gtl   
2       RL   Pave  SBrkr      IR1         Lvl    AllPub    Inside       Gtl   
3       RL   Pave  SBrkr      IR1         Lvl    AllPub    Corner       Gtl   
4       RL   Pave  SBrkr      IR1         Lvl    AllPub       FR2       Gtl   

  Neighborhood  Condition1  ... GarageType GarageFinish GarageQual GarageCond  \
0      CollgCr           1  ...     Attchd          RFn         TA         TA   
1      Veenker           2  ...     Attchd          RFn         TA         TA   
2      CollgCr           1  ...     Attchd          RFn         TA         TA   
3      Crawfor           1  ...     Detchd          Unf         TA         TA   
4      NoRidge           1  ...     Attchd          RFn         TA         TA   

  PavedDrive PoolQC  Fence MiscFeature SaleType SaleCondition  
0          Y  SBrkr  SBrkr       SBrkr       WD        Normal  
1          Y  SBrkr  SBrkr       SBrkr       WD        Normal  
2          Y  SBrkr  SBrkr       SBrkr       WD        Normal  
3          Y  SBrkr  SBrkr       SBrkr       WD       Abnorml  
4          Y  SBrkr  SBrkr       SBrkr       WD        Normal  

[5 rows x 43 columns]

### Label Encoding 

Another approach is to encode categorical values with a technique called "label encoding", which allows you to convert each value in a column to a number. Numerical labels are always between 0 and n_categories-1.

You can achieve the same label encoding using scikit-learn's LabelEncoder:

In [34]:
cat_df_sklearn = cat_df.copy()

from sklearn.preprocessing import LabelEncoder

lb = LabelEncoder()
cat_df_sklearn['Condition1'] = lb.fit_transform(cat_df['Condition1'])

cat_df_sklearn.head().iloc[:, 9:13]

Condition1 Condition2 BldgType HouseStyle
0           2       Norm     1Fam     2Story
1           1       Norm     1Fam     1Story
2           2       Norm     1Fam     2Story
3           2       Norm     1Fam     2Story
4           2       Norm     1Fam     2Story

Label encoding is very easy to grab but it has his pullbacks too. The numerical values can be misinterpreted by the machine learning algorithm. The algorithm takes numbers in hierarchy order.  

### One-Hot encoding 

What this does is that it converts each value in the category column(s) and turns it into new columns then it assigns a 1 or 0 (i.e. True/False) value to the columns. This make-up for the disadvantage in Label Encoding that attach hierarchy of importance to values.

This can be done with several libraries but the simplest way for doing this is the pandas' .get_dummies() method.

This pandas' .get_dummies() function is named this way because it creates dummy/indicator variables (1 or 0). There are mainly three arguments important here, the first one is the DataFrame you want to encode on, second being the columns argument which lets you specify the columns you want to do encoding on, and third, the prefix argument which lets you specify the prefix for the new columns that will be created after encoding.


In [18]:
cat_df_onehot = cat_df.copy()
cat_df_onehot = pd.get_dummies(cat_df_onehot, columns=['Condition1'], prefix = ['condition1'])

cat_df_onehot.head()

MSZoning Street  Alley LotShape LandContour Utilities LotConfig LandSlope  \
0       RL   Pave  SBrkr      Reg         Lvl    AllPub    Inside       Gtl   
1       RL   Pave  SBrkr      Reg         Lvl    AllPub       FR2       Gtl   
2       RL   Pave  SBrkr      IR1         Lvl    AllPub    Inside       Gtl   
3       RL   Pave  SBrkr      IR1         Lvl    AllPub    Corner       Gtl   
4       RL   Pave  SBrkr      IR1         Lvl    AllPub       FR2       Gtl   

  Neighborhood Condition2  ... SaleCondition condition1_Artery  \
0      CollgCr       Norm  ...        Normal                 0   
1      Veenker       Norm  ...        Normal                 0   
2      CollgCr       Norm  ...        Normal                 0   
3      Crawfor       Norm  ...       Abnorml                 0   
4      NoRidge       Norm  ...        Normal                 0   

  condition1_Feedr condition1_Norm condition1_PosA condition1_PosN  \
0                0               1               0               0   
1                1               0               0               0   
2                0               1               0               0   
3                0               1               0               0   
4                0               1               0               0   

  condition1_RRAe condition1_RRAn condition1_RRNe condition1_RRNn  
0               0               0               0               0  
1               0               0               0               0  
2               0               0               0               0  
3               0               0               0               0  
4               0               0               0               0  

[5 rows x 51 columns]

scikit-learn also supports one hot encoding through LabelBinarizer and OneHotEncoder in its preprocessing module. For this class we are going to do the same encoding through LabelBinarizer:

In [19]:
cat_df_onehot_sklearn = cat_df.copy()

from sklearn.preprocessing import LabelBinarizer

lb = LabelBinarizer()
lb_results = lb.fit_transform(cat_df_onehot_sklearn['Condition1'])
lb_results_df = pd.DataFrame(lb_results, columns=lb.classes_)

lb_results_df.head()

Artery  Feedr  Norm  PosA  PosN  RRAe  RRAn  RRNe  RRNn
0       0      0     1     0     0     0     0     0     0
1       0      1     0     0     0     0     0     0     0
2       0      0     1     0     0     0     0     0     0
3       0      0     1     0     0     0     0     0     0
4       0      0     1     0     0     0     0     0     0

Note that this lb_results_df resulted in a new DataFrame with only the one hot encodings for the feature Condition1. This needs to be concatenated back with the original DataFrame, which can be done via pandas' .concat() method. The axis argument is set to 1 as you want to merge on columns.

In [20]:
result_df = pd.concat([cat_df_onehot_sklearn, lb_results_df], axis=1)

result_df.head()

MSZoning Street  Alley LotShape LandContour Utilities LotConfig LandSlope  \
0       RL   Pave  SBrkr      Reg         Lvl    AllPub    Inside       Gtl   
1       RL   Pave  SBrkr      Reg         Lvl    AllPub       FR2       Gtl   
2       RL   Pave  SBrkr      IR1         Lvl    AllPub    Inside       Gtl   
3       RL   Pave  SBrkr      IR1         Lvl    AllPub    Corner       Gtl   
4       RL   Pave  SBrkr      IR1         Lvl    AllPub       FR2       Gtl   

  Neighborhood Condition1  ... SaleCondition Artery Feedr Norm PosA PosN RRAe  \
0      CollgCr       Norm  ...        Normal      0     0    1    0    0    0   
1      Veenker      Feedr  ...        Normal      0     1    0    0    0    0   
2      CollgCr       Norm  ...        Normal      0     0    1    0    0    0   
3      Crawfor       Norm  ...       Abnorml      0     0    1    0    0    0   
4      NoRidge       Norm  ...        Normal      0     0    1    0    0    0   

  RRAn RRNe RRNn  
0    0    0    0  
1    0    0    0  
2    0    0    0  
3    0    0    0  
4    0    0    0  

[5 rows x 52 columns]

While one-hot encoding solves the problem of unequal importance given to categories within a feature, it is not very useful when there are many categories, as that will result in formation of as many new columns, which can result in the curse of dimensionality. The concept of the “curse of dimensionality” discusses that in high-dimensional spaces some things just stop working properly.

### Binary Encoding

This technique is not as simple as the previous ones. In this technique, first the categories are encoded as ordinal, then those integers are converted into binary code, then the digits from that binary string are split into separate columns. This encodes the data in fewer dimensions than one-hot.

You can do binary encoding through a number of ways but the simplest one is using the category_encoders library. You can install category_encoders through pip install category_encoders on your command prompt, power shell or terminal.

You have to first import the category_encoders library after installing it. Invoke the BinaryEncoder function by specifying the columns you want to encode and then call the .fit_transform() method on it with the DataFrame as the argument.

In [29]:
cat_df_cenc = cat_df.copy()

import category_encoders as cenc

encoder = cenc.BinaryEncoder(cols=['Condition1'])
df_binary = encoder.fit_transform(cat_df_cenc)

df_binary.head().iloc[:, 10:]

Condition1_1  Condition1_2  Condition1_3  Condition1_4 Condition2 BldgType  \
0             0             0             0             1       Norm     1Fam   
1             0             0             1             0       Norm     1Fam   
2             0             0             0             1       Norm     1Fam   
3             0             0             0             1       Norm     1Fam   
4             0             0             0             1       Norm     1Fam   

  HouseStyle RoofStyle RoofMatl Exterior1st  ... GarageType GarageFinish  \
0     2Story     Gable  CompShg     VinylSd  ...     Attchd          RFn   
1     1Story     Gable  CompShg     MetalSd  ...     Attchd          RFn   
2     2Story     Gable  CompShg     VinylSd  ...     Attchd          RFn   
3     2Story     Gable  CompShg     Wd Sdng  ...     Detchd          Unf   
4     2Story     Gable  CompShg     VinylSd  ...     Attchd          RFn   

  GarageQual GarageCond PavedDrive PoolQC  Fence MiscFeature SaleType  \
0         TA         TA          Y  SBrkr  SBrkr       SBrkr       WD   
1         TA         TA          Y  SBrkr  SBrkr       SBrkr       WD   
2         TA         TA          Y  SBrkr  SBrkr       SBrkr       WD   
3         TA         TA          Y  SBrkr  SBrkr       SBrkr       WD   
4         TA         TA          Y  SBrkr  SBrkr       SBrkr       WD   

  SaleCondition  
0        Normal  
1        Normal  
2        Normal  
3       Abnorml  
4        Normal  

[5 rows x 37 columns]

Notice that four new columns are created in place of the carrier column with binary encoding for each category in the feature.

Note that category_encoders is a very useful library for encoding categorical columns. Not only does it support one-hot, binary and label encoding, but also other advanced encoding methods like Helmert contrast, polynomial contrast, backward difference, etc.

### Backward difference encoding

This technique falls under the contrast coding system for categorical features. A feature of K categories, or levels, usually enters a regression as a sequence of K-1 dummy variables. In backward difference coding, the mean of the dependent variable for a level is compared with the mean of the dependent variable for the prior level. This type of coding may be useful for a nominal or an ordinal variable.

The code structure is pretty much the same as any method in the category_encoders library, just this time you will call BackwardDifferenceEncoder from it:

In [28]:
encoder = cenc.BackwardDifferenceEncoder(cols=['Condition1'])
df_bdenc = encoder.fit_transform(cat_df_cenc)

df_bdenc.head().iloc[:, 10:]

Condition1_0  Condition1_1  Condition1_2  Condition1_3  Condition1_4  \
0     -0.888889     -0.777778     -0.666667     -0.555556     -0.444444   
1      0.111111     -0.777778     -0.666667     -0.555556     -0.444444   
2     -0.888889     -0.777778     -0.666667     -0.555556     -0.444444   
3     -0.888889     -0.777778     -0.666667     -0.555556     -0.444444   
4     -0.888889     -0.777778     -0.666667     -0.555556     -0.444444   

   Condition1_5  Condition1_6  Condition1_7 Condition2 BldgType  ...  \
0     -0.333333     -0.222222     -0.111111       Norm     1Fam  ...   
1     -0.333333     -0.222222     -0.111111       Norm     1Fam  ...   
2     -0.333333     -0.222222     -0.111111       Norm     1Fam  ...   
3     -0.333333     -0.222222     -0.111111       Norm     1Fam  ...   
4     -0.333333     -0.222222     -0.111111       Norm     1Fam  ...   

  GarageType GarageFinish GarageQual GarageCond PavedDrive PoolQC  Fence  \
0     Attchd          RFn         TA         TA          Y  SBrkr  SBrkr   
1     Attchd          RFn         TA         TA          Y  SBrkr  SBrkr   
2     Attchd          RFn         TA         TA          Y  SBrkr  SBrkr   
3     Detchd          Unf         TA         TA          Y  SBrkr  SBrkr   
4     Attchd          RFn         TA         TA          Y  SBrkr  SBrkr   

  MiscFeature SaleType SaleCondition  
0       SBrkr       WD        Normal  
1       SBrkr       WD        Normal  
2       SBrkr       WD        Normal  
3       SBrkr       WD       Abnorml  
4       SBrkr       WD        Normal  

[5 rows x 41 columns]

The interesting thing here is that you can see that the results are not the standard 1’s and 0’s you saw in the dummy encoding examples but rather regressed continuous values.

We have come to an end of the class. Stay blessed.